DEVHAS-731: Allow existing model servers to be passed into templates

johnmcollier opened the pull request:
Closes https://issues.redhat.com/browse/DEVHAS-731

This PR updates the AI software templates to allow existing model servers to be passed in when creating samples. When the `Existing model server` option is selected, the template will prompt the user to enter their model endpoint and model name values, which will then get passed into the gitops template (see also https://github.com/redhat-ai-dev/ai-lab-app/pull/20).

<img width="1235" alt="Screenshot 2024-07-30 at 5 18 57 PM" src="https://github.com/user-attachments/assets/92ec60fb-7a99-4839-90e6-6fe504743d27">

`generate.sh` also pulled in some base image updates to the Containerfiles, so I included them.

yangcao77 commented on skeleton/template.yaml:
```diff
@@ -45,8 +45,33 @@ spec:
+                modelEndpoint:
+                  title: Model Server Endpoint
+                  type: string
+                  ui:help: "The endpoint for an existing model server."
```
```suggestion
                  description: "The endpoint for an existing model server."
```

same as other descriptions

yangcao77 reviewed (commented):
Generally looks good. But I'm thinking if the existing model server option should restricted to be only for chatbot and codegen. 
for audio-to-text and object-detection, user can also provide an existing model server?

johnmcollier commented:
@yangcao77 I initially left it out of the audio-to-text sample since generate.sh didn't update it, but it was easy enough to add it in, so I included it. Note that the audio-to-text sample doesn't support specifying a model name, so I didn't include it in the template. 

The object detection template doesn't exist yet, but from looking at the dummy template.yaml that's generated when `generate.sh` runs, I see it's included, so it should be straightforward to include it there as well, once that template is done.

Please re-review when you get the chance, thanks!

yangcao77 commented:
> I initially left it out of the audio-to-text sample since generate.sh didn't update it
that's due to this
https://github.com/redhat-ai-dev/ai-lab-template/blob/fe674ce38f2e7bb200516f40a12e9d54c944429f/scripts/import-ai-lab-samples#L55-L57

since non-llm model templates did not support other model server option, so that model selection was disabled. 
and the `START` & `END` need to be updated to only exclude vllm option (probably rename it to be `SED_VLLM_MODEL_SERVER_START`/ `SED_VLLM_MODEL_SERVER_END`) https://github.com/redhat-ai-dev/ai-lab-template/blob/fe674ce38f2e7bb200516f40a12e9d54c944429f/skeleton/template.yaml#L40

> The object detection template doesn't exist yet
I just merged the PR include the object detection today afternoon. So you may want to rebase main to get the latest change. Sorry for the late notice

johnmcollier commented:
@yangcao77 I've updated the PR to support passing in models for all of the samples. I've done this by modifying the template skeleton to display the model server dropdown for all samples.

Because we don't want non-applicable model servers shown in the list for samples, I added more flags to the sample env files:
- `SUPPORT_VLLM` was changed to `SUPPORT_LLM`, as we don't want `llama.cpp` or `vllm` to be shown for non-llm samples
- `SUPPORT_ASR` was added for samples that require models that support automatic speech recognition (just `audio-to-text` currently)
- `SUPPORT_DETR` was added for samples that require models that support object recognition (just `object-detection` currently)

Additionally, when `SUPPORT_LLM` is set to true, we'll additionally prompt the user for a model name, as required by the current LLM samples we use.

yangcao77 reviewed (dismissed):
changes look good to me. thx

yangcao77 commented on templates/object-detection/template.yaml:
```diff
@@ -40,6 +40,7 @@ spec:
           title: DETR Model Server
           description: The model server container to use. The deployed model on the server must support object detection.
           type: string
+          default: detr-resnet-50
```
the default one we are using is detr-resnet-101

johnmcollier replied:
Fixed

johnmcollier commented:
@yangcao77 Please re-review, I had to fix a small bug with the default selection

yangcao77 reviewed (approved):
lgtm
